### PR TITLE
fix: add availability ton configure `authenticationFlowBindingOverrides` on client

### PR DESCRIPTION
**Pull request by jBouyoud**
This PR fix : https://github.com/adorsys/keycloak-config-cli/issues/170

### CODE REVIEW

**Comment by jkroepke** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java  22
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/adorsys/keycloak-config-cli/pullRequest?prid=6193725&bid=20207756)

**Comment by jkroepke** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java  22
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/adorsys/keycloak-config-cli/pullRequest?prid=6193725&bid=20207756)

**Comment by jkroepke** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java  22
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/adorsys/keycloak-config-cli/pullRequest?prid=6193725&bid=20207756)

**Comment by jkroepke** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java  27
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/adorsys/keycloak-config-cli/pullRequest?prid=6193725&bid=20207756)

**Comment by jkroepke** [comment]
> Native builds checks seems to be broken since : 1b134d3
Please do a rebase

File: src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java:76]: I think using deepClone here is a bit overkill since deepClone feels like a cost intensive operation.

authenticationFlowBindingOverrides could follow the same way like authorizationSettings. You could the that authorizationSettings is already mention on some DeepUtils calls (line 116, 127, 132) or does this not work?

File: src/test/java/de/adorsys/keycloak/config/AbstractImportTest.java
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/AbstractImportTest.java:68]: Any reason to be change this?

**Review by jkroepke — CHANGES_REQUESTED** [review]
Look good in generally. 👍 

I miss a test minimum with an id instead a name.

File: src/test/java/de/adorsys/keycloak/config/AbstractImportTest.java
Comment (jBouyoud) [reply src/test/java/de/adorsys/keycloak/config/AbstractImportTest.java:68]: Only for debugging purpose, WARN is not enough. 
I believe that could help the next guy to debug.
I can revert this if you want.

Comment (jkroepke) [reply src/test/java/de/adorsys/keycloak/config/AbstractImportTest.java:68]: All right. Since the runtime for integration tests are not changed, it's fine.

File: src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java
Comment (jBouyoud) [reply src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java]: I miss this. I'll fix that

**Comment by jkroepke** [comment]
> I miss a test minimum with an id instead a name.

Could you add a test that verifies that the import still works with a flow id?

File: src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java:386]: Is this required? 

You could set `authFlowUpdates` to an empty Map in line 375 and remove the condition in line 401+402 to just the else case.

Or does Keycloak require that all entries have to be an empty string?

**Comment by jkroepke** [comment]
If you not interested in some code optimization, just let me know and ignore my last comment. :)

Then I will merge it, if the test is there.

File: src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java
Comment (jBouyoud) [reply src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java:386]: Sadly; Keycloak require that all entries have to be an empty string in order to remove them

**Comment by jBouyoud** [comment]
As it, it's not possible to import a realm with an `authFlowBindingOverrides` with an id instead a name. Really sure that we need this ?

File: src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java
Comment (jkroepke) [reply src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java:386]: Ouhh... 

Could you has this Sadly information as commentary inside code? I will forget this in a few weeks... Thanks!

**Comment by jkroepke** [comment]
> As it, it's not possible to import a realm with an authFlowBindingOverrides with an id instead a name. Really sure that we need this ?

Are you 100% sure?

If I define the id of the flow inside my import like here

https://github.com/adorsys/keycloak-config-cli/blob/897155d2ed6d3d1ad233e08989ed0280550b6e38/src/test/resources/import-files/auth-flows/26_update_realm__update-non-top-level-flow-with-pseudo-id.json#L12

The id should be known, predictive and could be use by authFlowBindingOverrides.

Or does Keycloak override the id on creation?

Looking at Keycloaks source code, the id of the import model will be used if defined.

https://github.com/keycloak/keycloak/blob/5d5e56dde3bc841c7d1d0261cfcaf372cdf119bc/model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java#L1724

**Comment by jkroepke** [comment]
> As it, it's not possible to import a realm with an authFlowBindingOverrides with an id instead a name

Yes, it's possible. The example below partially work with master branch now but 2 runs are required first the flow, then the client. After that I could verify the the override setting at a client. Since the branch handle the import order of the settings, this error should be gone after merge.

Define overrides with an id works and are required if you want to import an exported realm with such configuration.

Demo realm config:
```
{
  "enabled": true,
  "realm": "test",
  "authenticationFlows": [
    {
      "id": "fbee9bfe-430a-48ac-8ef7-00dd17a1ab43",
      "alias": "custom flow",
      "description": "Custom flow for testing",
      "providerId": "basic-flow",
      "topLevel": true,
      "builtIn": false,
      "authenticationExecutions": [
        {
          "authenticator": "docker-http-basic-authenticator",
          "requirement": "DISABLED",
          "priority": 0,
          "userSetupAllowed": true,
          "autheticatorFlow": false
        }
      ]
    }
  ],
  "clients": [
    {
      "clientId": "another-client",
      "name": "another-client",
      "description": "Another-Client",
      "enabled": true,
      "clientAuthenticatorType": "client-secret",
      "secret": "my-other-client-secret",
      "redirectUris": [
        "*"
      ],
      "webOrigins": [
        "*"
      ],
      "authenticationFlowBindingOverrides": {
        "browser": "fbee9bfe-430a-48ac-8ef7-00dd17a1ab43"
      }
    }
  ]
}
```

Currently, the branch will throw an error that flow alias `fbee9bfe-430a-48ac-8ef7-00dd17a1ab43` could not be found. At review, I saw this potential issue, that was the reason for the requested test.

This is should should be fixed by lookup flows by id first and if none found, fallback to the flow alias lookup.

File: src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java:378]:
```suggestion
        // Be sure that all existing values will be cleared
        // See: https://github.com/keycloak/keycloak/blob/790b549cf99dbbba109e145654ee4a4cd1a047c9/server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java#L1516
```

**Comment by jBouyoud** [comment]
Hi, I'll try to add that no problem. 

When creating a new realm with a new flow, new client : no problem.
When updating a flow (in order to set an id, for exemple), since `keycloak-config-cli` recreates this flow and ignore `id` property ( https://github.com/adorsys/keycloak-config-cli/blob/2c0d021881b7641fa60c3c8667c7c22eef274c62/src/main/java/de/adorsys/keycloak/config/service/AuthenticationFlowsImportService.java#L266), is not possible to handle the case properly. 

I have no idea for why this property is ignored here, but import of `authenticationFlow` seems to be a little tricky. 
I'not sure about the impact of modifying this in `AuthenticationFlowsImportService`.

**Comment by jkroepke** [comment]
>  is not possible to handle the case properly.

I'm fine that this case just works on the initial realm creation.

A resource id feels immutable. If someone need this, he should create a new issue. 

As alternative, he could define an override by flow alias as this PR provide this.

**Comment by jBouyoud** [comment]
Shouldn't we handle only alias (on authFlow) to be more simple ?

**Comment by jkroepke** [comment]
I'm with you. 

But keycloak-config-cli should support a default realm export at minimum and could enrich experience in some points.

An export generated by keycloak itself would put an id here.

**Comment by jBouyoud** [comment]
Looks good to me

**Review by jkroepke — APPROVED** [review]
LGTM 👍

**Comment by ybonnefond** [comment]
@jkroepke, is it possible to release this?

**Comment by jkroepke** [comment]
I would like to take a look at #183 first.

If you are using the docker image, the tag master should include this fix.

**Comment by ybonnefond** [comment]
This is a critical piece of our system, I would rather not have our config set to master 😄 

Have you a rough idea how long it will take to address #183 ?

**Comment by jkroepke** [comment]
@ybonnefond 2.3.0 is out.

**Comment by ybonnefond** [comment]
Awesome, Thanks!

**Pull request merged**
